core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3955,11 +3960,490 @@ class KafkaApis(val requestChannel: RequestChannel,
+            leaderNode.node.foreach { node =>
+              nodeEndpoints.put(node.id, node)
+            }
+            partitionData.currentLeader
+              .setLeaderId(leaderNode.leaderId)
+              .setLeaderEpoch(leaderNode.leaderEpoch)
+          case _ =>
+        }
+
+        partitions.put(tp, partitionData)
+      }
+      erroneous.foreach { case (tp, data) => partitions.put(tp, data) }
+
+      var unconvertedShareFetchResponse: ShareFetchResponse = null
+
+      def createResponse(throttleTimeMs: Int): ShareFetchResponse = {
+        // Down-convert messages for each partition if required
+        val convertedData = new util.LinkedHashMap[TopicIdPartition, ShareFetchResponseData.PartitionData]
+        unconvertedShareFetchResponse.data.responses.forEach { topicResponse =>
+          topicResponse.partitions.forEach { unconvertedPartitionData =>
+            val tp = new TopicIdPartition(topicResponse.topicId, new TopicPartition(topicNames.get(topicResponse.topicId),
+              unconvertedPartitionData.partitionIndex))
+            val error = Errors.forCode(unconvertedPartitionData.errorCode)
+            if (error != Errors.NONE)
+              debug(s"Share Fetch request with correlation id ${request.header.correlationId} from client $clientId " +
+                s"on partition $tp failed due to ${error.exceptionName}")
+            convertedData.put(tp, maybeConvertShareFetchedData(tp, unconvertedPartitionData))
+          }
+        }
+
+        // Prepare share fetch response from converted data
+        val response =
+          ShareFetchResponse.of(unconvertedShareFetchResponse.error, throttleTimeMs, convertedData, nodeEndpoints.values.toList.asJava)
+        // record the bytes out metrics only when the response is being sent
+        response.data.responses.forEach { topicResponse =>
+          topicResponse.partitions.forEach { data =>
+            // If the topic name was not known, we will have no bytes out.
+            if (topicResponse.topicId != null) {
+              val tp = new TopicIdPartition(topicResponse.topicId, new TopicPartition(topicNames.get(topicResponse.topicId), data.partitionIndex))
+              brokerTopicStats.updateBytesOut(tp.topic, false, false, ShareFetchResponse.recordsSize(data))
+            }
+          }
+        }
+        response
+      }
+      // Share Fetch size used to determine throttle time is calculated before any down conversions.
+      // This may be slightly different from the actual response size. But since down conversions
+      // result in data being loaded into memory, we should do this only when we are not going to throttle.
+      //
+      // Record both bandwidth and request quota-specific values and throttle by muting the channel if any of the
+      // quotas have been violated. If both quotas have been violated, use the max throttle time between the two
+      // quotas. When throttled, we unrecord the recorded bandwidth quota value
```

Comment:
nit: Add a "." at the end of comment. Can you do this for other added comments as well?